### PR TITLE
mgr/cephadm: Add GrafanaSpec.initial_admin_password

### PR DESCRIPTION
**Pull request by sebastian-philipp**
By default, we're not creating any admin accout for Grafana now,
but we're adding an option to set the grafana password manually using:

```yaml
service_type: grafana
spec:
  initial_admin_password: mypassword
```

Users can then easily log into Grafana with the given password.

Fixes: https://tracker.ceph.com/issues/48291
Signed-off-by: Sebastian Wagner <sewagner@redhat.com>

image

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by p-se** [comment]
I think disabling the admin account by default is a sane decision, especially considering that the admin account is not required by Ceph Dashboard to embed the Grafana graphs.

However, having to replace the whole Grafana configuration file just to enable the admin account, requires us to keep the `src/pybind/mgr/cephadm/templates/services/grafana/grafana.ini.j2` file in sync with the documentation, to not introduce any bugs. It (currently) would also prevent updates of the Grafana configuration file to reach the user, as the user-defined template is not automatically replaced or merged by/with the automatically generated configuration file, but stays enabled until the user decides to change it (and we do not inform the user about any updates yet, as far as I know). I therefore think this might be good opportunity to introduce another cephadm setting which is evaluated in the Grafana Jinja Template and enabled/disables the admin user.

**Comment by sebastian-philipp** [comment]
> I think disabling the admin account by default is a reasonable decision, especially considering that the admin account is not required by Ceph Dashboard to embed the Grafana graphs.
> 
> However, having to replace the whole Grafana configuration file just to enable the admin account, requires us to keep the `src/pybind/mgr/cephadm/templates/services/grafana/grafana.ini.j2` file in sync with the documentation, to not introduce any bugs. It (currently) would also prevent updates of the Grafana configuration file to reach the user, as the user-defined template is not automatically replaced or merged by/with the automatically generated configuration file, but stays enabled until the user decides to change it (and we do not inform the user about any updates yet, as far as I know). 

Completely agree here. *But*

> I therefore think this might be good opportunity to introduce another cephadm setting which is evaluated in the Grafana Jinja Template and enabled/disables the admin user.

*But*, we're not going to prevent this. If we go with this PR here, we'll have two options in the future:

1. Provide a better way to deal with updates of default templates. (three-way-diff, notify users, etc)
2. Introduce new options, switches and flags to the specs for things users really demand. 

I'd actually go with this PR here and revisit this later on.

**Comment by sebastian-philipp** [comment]
Ok turns out the cli is no use here:
```
bash-4.4$ grafana-cli admin reset-admin-password foobar
INFO[07-22|11:37:14] Connecting to DB                         logger=sqlstore dbtype=sqlite3
INFO[07-22|11:37:14] Starting DB migration                    logger=migrator
INFO[07-22|11:37:14] Created default organization             logger=sqlstore
Error: ✗ could not read user from database. Error: User not found
bash-4.4$
```

but setting the template is works.

**Comment by sebastian-philipp** [comment]
/home/jenkins-build/build/workspace/ceph-pr-docs/doc/cephadm/monitoring.rst:378: WARNING: Mismatch: both interpreted text role prefix and reference suffix.

File: doc/cephadm/monitoring.rst [hunk 1]
```diff
@@ -337,3 +339,44 @@ Enabling RBD-Image monitoring
...
+
+Then apply it::
+
+  ceph config-key set mgr/cephadm/services/grafana/grafana.ini -i grafana.ini
+  ceph orch redeploy grafana
+
+.. note::
+
+  Overwriting the Jinja2 template has a downside. See
+  :ref:`cephadm-overwrite-jinja2-templates`_. Also, feel free to add a note to
```
Comment (mgfritch) [inline doc/cephadm/monitoring.rst:379]:
```suggestion
  :ref:`cephadm-overwrite-jinja2-templates`. Also, feel free to add a note to
```

**Review by mgfritch — APPROVED** [review]

**Comment by epuertat** [comment]
> I think disabling the admin account by default is a reasonable decision, especially considering that the admin account is not required by Ceph Dashboard to embed the Grafana graphs.
> 
> However, having to replace the whole Grafana configuration file just to enable the admin account, requires us to keep the `src/pybind/mgr/cephadm/templates/services/grafana/grafana.ini.j2` file in sync with the documentation, to not introduce any bugs. It (currently) would also prevent updates of the Grafana configuration file to reach the user, as the user-defined template is not automatically replaced or merged by/with the automatically generated configuration file, but stays enabled until the user decides to change it (and we do not inform the user about any updates yet, as far as I know). I therefore think this might be good opportunity to introduce another cephadm setting which is evaluated in the Grafana Jinja Template and enabled/disables the admin user.

@p-se I mentioned exactly the same thing to @sebastian-philipp. IMHO we should question whether to add or not more knobs to cephadm when they don't have a clear user impact, but in this very case we do know of users accessing and modifying the Grafana dashboards (that's also why we are mounting and persisting the Grafana DB directory). We may argue that this will only affect new deployments, but that doesn't change that fact that this might be perceived as a regression.

Can't we add a new property to the Grafana spec to deal with this case?

**Comment by pcuzner** [comment]
@sebastian-philipp What's the motivation behind this change? The PR and commit is not describing anything?

The data collected by prometheus belongs to the user/customer - so why would we block them being able to use grafana tro visualise their data? For example, without admin I couldn't look at adding capacity planning, or include any insights from cadvisor? The dashboards we provide are provisioned dashboards, so they're not immediately editable by the user anyway - so I don't see that as an exposure.

Why wouldn't we just set/accept an admin password to avoid unnecessary exposure of grafana(tighten security) and avoid regression as per @epuertat comment?

**Comment by sebastian-philipp** [comment]
@p-se  @epuertat and @pcuzner : Fair enough. I added a switch to set the initial admin password. Does this work for you?

File: src/pybind/mgr/cephadm/templates/services/grafana/grafana.ini.j2 [hunk 2]
```diff
@@ -13,6 +13,10 @@
   http_port = {{ http_port }}
   http_addr = {{ http_addr }}
 [security]
+{% if not initial_admin_password %}
+  disable_initial_admin_creation = true
+{% else %}
   admin_user = admin
-  admin_password = admin
+  admin_password = {{ initial_admin_password }}
+{% endif %}
```
Comment (epuertat) [inline src/pybind/mgr/cephadm/templates/services/grafana/grafana.ini.j2:21]: Thanks a lot!

File: src/pybind/mgr/cephadm/tests/test_services.py [hunk 3]
```diff
@@ -223,3 +227,46 @@ def test_iscsi_dashboard_config(self, mock_resolve_ip):
                                        'https://user:password@[FEDC:BA98:7654:3210:FEDC:BA98:7654:3210]:5000')
 
         assert dashboard_expected_call in self.mgr.check_mon_command.mock_calls
+
+
+class TestGrafanaService:
+    @mock.patch("cephadm.serve.CephadmServe._run_cephadm", _run_cephadm('{}'))
+    def testgenerate_config(self, cephadm_module: CephadmOrchestrator):
+        with with_host(cephadm_module, 'test'):
+            with with_service(cephadm_module, GrafanaSpec(initial_admin_password='secure')):
```
Comment (epuertat) [inline src/pybind/mgr/cephadm/tests/test_services.py:236]: You can have multiple context managers in a single with statement:
```suggestion
        with with_host(cephadm_module, 'test'), with_service(cephadm_module, GrafanaSpec(initial_admin_password='secure')):
```

**Review by epuertat — APPROVED** [review]

File: src/pybind/mgr/cephadm/tests/test_services.py (hunk 3 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/tests/test_services.py:236]: Right. This is super helpful, if you have lots of mock patches for example:

https://github.com/ceph/ceph/blob/ade690b549c57d0cab8903227bdf984c3772b774/src/cephadm/tests/fixtures.py#L50-L53

**Review by jmolmo — APPROVED** [review]

File: src/pybind/mgr/cephadm/tests/test_services.py (hunk 3 above)
Comment (epuertat) [reply src/pybind/mgr/cephadm/tests/test_services.py:236]: For that you could use `patch.multiple(cephamd.CephadmDaemon, daemon_path=_daemon_path(), can_run=Mock(return_value=True), ...)`

**Comment by tchaikov** [comment]
```
2021-08-05T05:10:27.840 INFO:journalctl@ceph.mon.b.smithi087.stdout:Aug 05 05:10:27 smithi087 ceph-mon[29866]: from='mgr.14150 172.21.15.53:0/2007501462' entity='mgr.y' cmd=[{"prefix":"config rm","who":"mgr","name":"mgr/rbd_support/y/trash_purge_schedule"}]: dispatch
2021-08-05T05:10:27.840 INFO:journalctl@ceph.mon.b.smithi087.stdout:Aug 05 05:10:27 smithi087 ceph-mon[29866]: from='mgr.14150 172.21.15.53:0/2007501462' entity='mgr.y' cmd=[{"prefix":"config rm","who":"mgr","name":"mgr/rbd_support/y/mirror_snapshot_schedule"}]: dispatch
2021-08-05T05:10:28.353 DEBUG:teuthology.orchestra.run.smithi053:> sudo /home/ubuntu/cephtest/cephadm --image quay.ceph.io/ceph-ci/ceph:35b9397fa260b8fbb2d83a5d1200e6a289d5d74d shell -c /etc/ceph/ceph.conf -k /etc/ceph/ceph.client.admin.keyring --fsid 9911b090-f5a9-11eb-8c24-001a4aab830c -- bash -c 'ceph orch ls --format yaml'
2021-08-05T05:10:28.568 INFO:journalctl@ceph.mon.a.smithi053.stdout:Aug 05 05:10:28 smithi053 ceph-mon[28793]: pgmap v326: 132 pgs: 132 active+clean; 582 KiB data, 52 MiB used, 715 GiB / 715 GiB avail; 853 B/s rd, 0 op/s
2021-08-05T05:10:28.569 INFO:journalctl@ceph.mon.a.smithi053.stdout:Aug 05 05:10:28 smithi053 ceph-mon[28793]: from='client.24610 -' entity='client.admin' cmd=[{"prefix": "orch device ls", "target": ["mon-mgr", ""]}]: dispatch
2021-08-05T05:10:28.569 INFO:journalctl@ceph.mon.c.smithi053.stdout:Aug 05 05:10:28 smithi053 ceph-mon[35169]: pgmap v326: 132 pgs: 132 active+clean; 582 KiB data, 52 MiB used, 715 GiB / 715 GiB avail; 853 B/s rd, 0 op/s
2021-08-05T05:10:28.569 INFO:journalctl@ceph.mon.c.smithi053.stdout:Aug 05 05:10:28 smithi053 ceph-mon[35169]: from='client.24610 -' entity='client.admin' cmd=[{"prefix": "orch device ls", "target": ["mon-mgr", ""]}]: dispatch
2021-08-05T05:10:28.840 INFO:journalctl@ceph.mon.b.smithi087.stdout:Aug 05 05:10:28 smithi087 ceph-mon[29866]: pgmap v326: 132 pgs: 132 active+clean; 582 KiB data, 52 MiB used, 715 GiB / 715 GiB avail; 853 B/s rd, 0 op/s
2021-08-05T05:10:28.841 INFO:journalctl@ceph.mon.b.smithi087.stdout:Aug 05 05:10:28 smithi087 ceph-mon[29866]: from='client.24610 -' entity='client.admin' cmd=[{"prefix": "orch device ls", "target": ["mon-mgr", ""]}]: dispatch
2021-08-05T05:10:30.577 INFO:teuthology.orchestra.run.smithi053.stderr:Error EINVAL: Failed to validate Drive Group: OSD spec needs a `placement` key.
2021-08-05T05:10:30.840 INFO:journalctl@ceph.mon.b.smithi087.stdout:Aug 05 05:10:30 smithi087 ceph-mon[29866]: pgmap v327: 132 pgs: 132 active+clean; 582 KiB data, 52 MiB used, 715 GiB / 715 GiB avail; 853 B/s rd, 0 op/s
2021-08-05T05:10:30.901 INFO:journalctl@ceph.mon.a.smithi053.stdout:Aug 05 05:10:30 smithi053 ceph-mon[28793]: pgmap v327: 132 pgs: 132 active+clean; 582 KiB data, 52 MiB used, 715 GiB / 715 GiB avail; 853 B/s rd, 0 op/s
2021-08-05T05:10:30.902 INFO:journalctl@ceph.mgr.y.smithi053.stdout:Aug 05 05:10:30 smithi053 conmon[29110]: 2021-08-05T05:10:30.571+0000 7f3e70bd2700 -1 mgr.server reply reply (22) Invalid argument Failed to validate Drive Group: OSD spec needs a `placement` key.
2021-08-05T05:10:30.902 INFO:journalctl@ceph.mon.c.smithi053.stdout:Aug 05 05:10:30 smithi053 ceph-mon[35169]: pgmap v327: 132 pgs: 132 active+clean; 582 KiB data, 52 MiB used, 715 GiB / 715 GiB avail; 853 B/s rd, 0 op/s
2021-08-05T05:10:31.288 DEBUG:teuthology.orchestra.run:got remote process result: 22
2021-08-05T05:10:31.289 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_04c2febe7099917d97a71271f17abb5710030132/teuthology/run_tasks.py", line 91, in run_tasks
    manager = run_one_task(taskname, ctx=ctx, config=config)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_04c2febe7099917d97a71271f17abb5710030132/teuthology/run_tasks.py", line 70, in run_one_task
    return task(**kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_35b9397fa260b8fbb2d83a5d1200e6a289d5d74d/qa/tasks/cephadm.py", line 1034, in shell
    extra_cephadm_args=args)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_35b9397fa260b8fbb2d83a5d1200e6a289d5d74d/qa/tasks/cephadm.py", line 46, in _shell
    **kwargs
```

**Comment by epuertat** [comment]
jenkins test dashboard

**Comment by epuertat** [comment]
Dashboard supports disabling the anonymous mode with the following changes to grafana.ini: https://github.com/rhcs-dashboard/ceph-dev/pull/46/files#diff-4e718ad62eacf3cfd51eb1b417dd8d93ba7954563beef2e468908eb4af8d9616

File: src/pybind/mgr/cephadm/templates/services/grafana/grafana.ini.j2 [hunk 4]
```diff
@@ -13,6 +13,10 @@
...
   allow_embedding = true
```
Comment (epuertat) [inline src/pybind/mgr/cephadm/templates/services/grafana/grafana.ini.j2:22]: BTW, according to the [PR I linked](https://github.com/rhcs-dashboard/ceph-dev/pull/46/files#), we need to add these settings under `[security]` section:
```
cookie_secure = true
cookie_samesite = none
```

And optionally disable the anonymous mode:
```
[auth.anonymous]
enabled = {{ grafana_anonymous_mode }}
```

**Review by epuertat — CHANGES_REQUESTED** [review]

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Comment by sebastian-philipp** [comment]
jenkins test dashboard

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2021-11-29_08:27:28-orch:cephadm-wip-swagner-testing-2021-11-26-1656-distro-default-smithi/

https://tracker.ceph.com/issues/53424

**Review by epuertat — APPROVED** [review]

**Pull request merged**